pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,240 @@
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get(); // Just make sure it's initialized
+    }
+    
+    /**
+     * Finishes tracking a thread.
+     */
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        // if using a mono-thread, we may not be empty...
+        if (TIMER_ENTRIES.get().size() == 1) {
+            finishOperation();
+            TIMER_ENTRIES.remove();
+        }
+    }
+    
+    /**
+     * Starts tracking an operation
+     * @param category The category under which to track the operation.
+     */
+    public static void startOperation(final TimedOperationCategory category) {
+        startOperation(category, null);
+    }
+    
+    /**
+     * Starts tracking an operation
+     * @param category The category under which to track the operation.
+     * @param label A label to be added to the category. Allows to differentiate measures within a single category.
+     */
+    public static void startOperation(final TimedOperationCategory category, final String label) {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get().add(new TimerEntry(category, label));
+    }
+    
+    /**
+     * Finishes tracking an operation
+     */
+    public static void finishOperation() {
+        finishOperation(0);
+    }
+    
+    /**
+     * Finishes tracking an operation
+     * @param extraDataCounter An optional additional data counter to track along the measurements
+     */
+    public static void finishOperation(final long extraDataCounter) {
+        if (!trackTime) {
+            return;
+        }
+
+        final Queue<TimerEntry> queue = TIMER_ENTRIES.get();
+        final TimerEntry timerEntry = queue.remove();
+
+        // Compute if absent
+        TimedResult result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        if (result == null) {
+            ACCUMULATED_RESULTS.putIfAbsent(timerEntry.operation, new TimedResult());
+            result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        }
+
+        // Update counters and let next element on the stack ignore the time we spent
+        final long delta = result.accumulate(timerEntry, extraDataCounter);
+        if (!queue.isEmpty()) { 
+            queue.peek().inNestedOperations += delta;
+        }
+    }
+    
+    private static class TimerEntry {
+        /* package */ final TimedOperation operation;
+        /* package */ final long start;
+        /* package */ long inNestedOperations = 0;
+        
+        /* package */ TimerEntry(final TimedOperationCategory category, final String label) {
+            this.operation = new TimedOperation(category, label);
+            this.start = System.nanoTime();
+        }
+
+        @Override
+        public String toString() {
+            return "TimerEntry for " + operation;
+        }
+    }
+    
+    /* package */ static class TimedResult {
```

Comment:
I'd like a javadoc comment on that. I know it's not important for the public API javadocs, but it would make the code easier to grok